File: rust-example/src/capabilities/book_stay/sql.rs
use chrono::NaiveDate;
use sqlx::{Postgres, Row, Transaction};

use super::context::{
    BookingContext, GuestBookingEligibility, ListingBookingSettings, ListingBookingStatus,
};
use super::request::BookStay;
use super::result::ReservationConfirmed;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct LoadedBookingState {
    pub guest: Option<GuestBookingEligibility>,
    pub listing: Option<ListingBookingSettings>,
    pub unavailable_nights: Vec<NaiveDate>,
}

impl LoadedBookingState {
    pub fn into_context(self) -> BookingContext {
        BookingContext {
            guest: self.guest,
            listing: self.listing,
            unavailable_nights: self.unavailable_nights,
        }
    }
}

#[derive(Debug, thiserror::Error)]
pub enum LoadBookingStateError {
    #[error("database error: {0}")]
    Database(#[from] sqlx::Error),

    #[error("invalid stored value: {0}")]
    InvalidStoredValue(String),
}

#[derive(Debug, thiserror::Error)]
pub enum RecordReservationError {
    #[error("database error: {0}")]
    Database(sqlx::Error),

    #[error("invalid stored value: {0}")]
    InvalidStoredValue(String),

    #[error("listing unavailable")]
    ListingUnavailable,
}

pub async fn load_booking_state(
    tx: &mut Transaction<'_, Postgres>,
    request: &BookStay,
) -> Result<LoadedBookingState, LoadBookingStateError> {
    let guest = load_guest_eligibility(tx, request).await?;
    let listing = load_listing_booking_settings(tx, request).await?;
    let unavailable_nights = load_unavailable_nights(tx, request).await?;

    Ok(LoadedBookingState {
        guest,
        listing,
        unavailable_nights,
    })
}

async fn load_guest_eligibility(
    tx: &mut Transaction<'_, Postgres>,
    request: &BookStay,
) -> Result<Option<GuestBookingEligibility>, LoadBookingStateError> {
    let row = sqlx::query(
        r#"
        SELECT booking_eligibility
        FROM guests
        WHERE id = $1
        FOR SHARE
        "#,
    )
    .bind(request.guest_id)
    .fetch_optional(&mut **tx)
    .await?;

    row.map(|row| {
        let value: String = row.get("booking_eligibility");
        match value.as_str() {
            "eligible" => Ok(GuestBookingEligibility::Eligible),
            "blocked" => Ok(GuestBookingEligibility::Blocked),
            _ => Err(LoadBookingStateError::InvalidStoredValue(format!(
                "unknown guest booking eligibility: {value}"
            ))),
        }
    })
    .transpose()
}

async fn load_listing_booking_settings(
    tx: &mut Transaction<'_, Postgres>,
    request: &BookStay,
) -> Result<Option<ListingBookingSettings>, LoadBookingStateError> {
    let row = sqlx::query(
        r#"
        SELECT booking_status, max_guests, min_nights, max_nights
        FROM listings
        WHERE id = $1
        FOR SHARE
        "#,
    )
    .bind(request.listing_id)
    .fetch_optional(&mut **tx)
    .await?;

    row.map(|row| {
        let booking_status: String = row.get("booking_status");
        let max_guests: i32 = row.get("max_guests");
        let min_nights: i32 = row.get("min_nights");
        let max_nights: Option<i32> = row.get("max_nights");

        Ok(ListingBookingSettings {
            booking_status: map_listing_booking_status(&booking_status)?,
            max_guests: i32_to_u32("listings.max_guests", max_guests)?,
            min_nights: i32_to_u32("listings.min_nights", min_nights)?,
            max_nights: max_nights
                .map(|value| i32_to_u32("listings.max_nights", value))
                .transpose()?,
        })
    })
    .transpose()
}

async fn load_unavailable_nights(
    tx: &mut Transaction<'_, Postgres>,
    request: &BookStay,
) -> Result<Vec<NaiveDate>, LoadBookingStateError> {
    let rows = sqlx::query(
        r#"
        SELECT night
        FROM listing_unavailable_nights
        WHERE listing_id = $1
          AND night >= $2
          AND night < $3
        ORDER BY night
        "#,
    )
    .bind(request.listing_id)
    .bind(request.stay.check_in)
    .bind(request.stay.check_out)
    .fetch_all(&mut **tx)
    .await?;

    Ok(rows.into_iter().map(|row| row.get("night")).collect())
}

pub async fn record_reservation_confirmed(
    tx: &mut Transaction<'_, Postgres>,
    confirmed: &ReservationConfirmed,
) -> Result<(), RecordReservationError> {
    let guest_count = i32::try_from(confirmed.guest_count).map_err(|_| {
        RecordReservationError::InvalidStoredValue(
            "guest_count exceeds PostgreSQL INTEGER".to_owned(),
        )
    })?;
    let max_guests_at_confirmation = u32_to_i32(
        "max_guests_at_confirmation",
        confirmed.max_guests_at_confirmation,
    )?;
    let min_nights_at_confirmation = u32_to_i32(
        "min_nights_at_confirmation",
        confirmed.min_nights_at_confirmation,
    )?;
    let max_nights_at_confirmation = confirmed
        .max_nights_at_confirmation
        .map(|value| u32_to_i32("max_nights_at_confirmation", value))
        .transpose()?;

    sqlx::query(
        r#"
        INSERT INTO reservations (
            id,
            listing_id,
            guest_id,
            check_in,
            check_out,
            guest_count,
            status,
            confirmed_at,
            max_guests_at_confirmation,
            min_nights_at_confirmation,
            max_nights_at_confirmation
        )
        VALUES ($1, $2, $3, $4, $5, $6, 'confirmed', $7, $8, $9, $10)
        "#,
    )
    .bind(confirmed.reservation_id)
    .bind(confirmed.listing_id)
    .bind(confirmed.guest_id)
    .bind(confirmed.stay.check_in)
    .bind(confirmed.stay.check_out)
    .bind(guest_count)
    .bind(confirmed.confirmed_at)
    .bind(max_guests_at_confirmation)
    .bind(min_nights_at_confirmation)
    .bind(max_nights_at_confirmation)
    .execute(&mut **tx)
    .await
    .map_err(RecordReservationError::Database)?;

    let mut night = confirmed.stay.check_in;

    while night < confirmed.stay.check_out {
        sqlx::query(
            r#"
            INSERT INTO listing_unavailable_nights (
                listing_id,
                night,
                unavailability_type,
                reservation_id
            )
            VALUES ($1, $2, 'reservation', $3)
            "#,
        )
        .bind(confirmed.listing_id)
        .bind(night)
        .bind(confirmed.reservation_id)
        .execute(&mut **tx)
        .await
        .map_err(|error| {
            if is_unique_violation(&error) {
                RecordReservationError::ListingUnavailable
            } else {
                RecordReservationError::Database(error)
            }
        })?;

        night = night
            .succ_opt()
            .expect("date range should not overflow after validation");
    }

    Ok(())
}

fn map_listing_booking_status(value: &str) -> Result<ListingBookingStatus, LoadBookingStateError> {
    match value {
        "bookable" => Ok(ListingBookingStatus::Bookable),
        "disabled" => Ok(ListingBookingStatus::Disabled),
        _ => Err(LoadBookingStateError::InvalidStoredValue(format!(
            "unknown listing booking status: {value}"
        ))),
    }
}

fn i32_to_u32(field: &str, value: i32) -> Result<u32, LoadBookingStateError> {
    u32::try_from(value).map_err(|_| {
        LoadBookingStateError::InvalidStoredValue(format!(
            "{field} must be representable as u32: {value}"
        ))
    })
}

fn u32_to_i32(field: &str, value: u32) -> Result<i32, RecordReservationError> {
    i32::try_from(value).map_err(|_| {
        RecordReservationError::InvalidStoredValue(format!(
            "{field} exceeds PostgreSQL INTEGER: {value}"
        ))
    })
}

fn is_unique_violation(error: &sqlx::Error) -> bool {
    error
        .as_database_error()
        .and_then(|database_error| database_error.code())
        .is_some_and(|code| code == "23505")
}

File: csharp-example/src/BookingAStay/Capabilities/BookStay/Sql.cs
using Dapper;
using Npgsql;

namespace BookingAStay.Capabilities.BookStay;

public sealed record LoadedBookingState(
    GuestBookingEligibility? Guest,
    ListingBookingSettings? Listing,
    IReadOnlyList<DateOnly> UnavailableNights)
{
    public BookingContext ToContext()
    {
        return new BookingContext(Guest, Listing, UnavailableNights);
    }
}

public sealed class InvalidStoredValueException(string message) : Exception(message);

public sealed class ListingUnavailableException : Exception
{
}

public static class Sql
{
    internal sealed record GuestRow(string BookingEligibility);

    internal sealed record ListingRow(
        string BookingStatus,
        int MaxGuests,
        int MinNights,
        int? MaxNights);

    internal sealed record UnavailableNightRow(DateOnly Night);

    public static async Task<LoadedBookingState> LoadBookingState(
        NpgsqlConnection connection,
        NpgsqlTransaction transaction,
        BookStayRequest request,
        CancellationToken cancellationToken = default)
    {
        var guest = await LoadGuestEligibility(connection, transaction, request, cancellationToken);
        var listing = await LoadListingBookingSettings(connection, transaction, request, cancellationToken);
        var unavailableNights = await LoadUnavailableNights(connection, transaction, request, cancellationToken);

        return new LoadedBookingState(guest, listing, unavailableNights);
    }

    public static async Task RecordReservationConfirmed(
        NpgsqlConnection connection,
        NpgsqlTransaction transaction,
        ReservationConfirmed confirmed,
        CancellationToken cancellationToken = default)
    {
        const string reservationSql = """
            INSERT INTO reservations (
                id,
                listing_id,
                guest_id,
                check_in,
                check_out,
                guest_count,
                status,
                confirmed_at,
                max_guests_at_confirmation,
                min_nights_at_confirmation,
                max_nights_at_confirmation
            )
            VALUES (@id, @listing_id, @guest_id, @check_in, @check_out, @guest_count, 'confirmed', @confirmed_at, @max_guests, @min_nights, @max_nights)
            """;

        await connection.ExecuteAsync(
            reservationSql,
            new
            {
                id = confirmed.ReservationId,
                listing_id = confirmed.ListingId,
                guest_id = confirmed.GuestId,
                check_in = confirmed.Stay.CheckIn,
                check_out = confirmed.Stay.CheckOut,
                guest_count = confirmed.GuestCount,
                confirmed_at = confirmed.ConfirmedAt,
                max_guests = confirmed.MaxGuestsAtConfirmation,
                min_nights = confirmed.MinNightsAtConfirmation,
                max_nights = confirmed.MaxNightsAtConfirmation,
                cancellationToken,
            },
            transaction);

        for (var night = confirmed.Stay.CheckIn; night < confirmed.Stay.CheckOut; night = night.AddDays(1))
        {
            const string unavailableNightSql = """
                INSERT INTO listing_unavailable_nights (
                    listing_id,
                    night,
                    unavailability_type,
                    reservation_id
                )
                VALUES (@listing_id, @night, 'reservation', @reservation_id)
                """;

            try
            {
                await connection.ExecuteAsync(
                    unavailableNightSql,
                    new
                    {
                        listing_id = confirmed.ListingId,
                        night,
                        reservation_id = confirmed.ReservationId,
                        cancellationToken,
                    },
                    transaction);
            }
            catch (PostgresException exception) when (exception.SqlState == PostgresErrorCodes.UniqueViolation)
            {
                throw new ListingUnavailableException();
            }
        }
    }

    private static async Task<GuestBookingEligibility?> LoadGuestEligibility(
        NpgsqlConnection connection,
        NpgsqlTransaction transaction,
        BookStayRequest request,
        CancellationToken cancellationToken)
    {
        const string sql = """
            SELECT booking_eligibility AS BookingEligibility
            FROM guests
            WHERE id = @guest_id
            FOR SHARE
            """;

        var row = await connection.QuerySingleOrDefaultAsync<GuestRow>(
            sql,
            new { guest_id = request.GuestId, cancellationToken },
            transaction);
        var value = row?.BookingEligibility;

        return value switch
        {
            null => null,
            "eligible" => GuestBookingEligibility.Eligible,
            "blocked" => GuestBookingEligibility.Blocked,
            _ => throw new InvalidStoredValueException($"unknown guest booking eligibility: {value}"),
        };
    }

    private static async Task<ListingBookingSettings?> LoadListingBookingSettings(
        NpgsqlConnection connection,
        NpgsqlTransaction transaction,
        BookStayRequest request,
        CancellationToken cancellationToken)
    {
        const string sql = """
            SELECT
                booking_status AS BookingStatus,
                max_guests AS MaxGuests,
                min_nights AS MinNights,
                max_nights AS MaxNights
            FROM listings
            WHERE id = @listing_id
            FOR SHARE
            """;

        var row = await connection.QuerySingleOrDefaultAsync<ListingRow>(
            sql,
            new { listing_id = request.ListingId, cancellationToken },
            transaction);

        if (row is null)
        {
            return null;
        }

        var bookingStatus = row.BookingStatus switch
        {
            "bookable" => ListingBookingStatus.Bookable,
            "disabled" => ListingBookingStatus.Disabled,
            var value => throw new InvalidStoredValueException($"unknown listing booking status: {value}"),
        };

        return new ListingBookingSettings(
            bookingStatus,
            row.MaxGuests,
            row.MinNights,
            row.MaxNights);
    }

    private static async Task<IReadOnlyList<DateOnly>> LoadUnavailableNights(
        NpgsqlConnection connection,
        NpgsqlTransaction transaction,
        BookStayRequest request,
        CancellationToken cancellationToken)
    {
        const string sql = """
            SELECT night AS Night
            FROM listing_unavailable_nights
            WHERE listing_id = @listing_id
              AND night >= @check_in
              AND night < @check_out
            ORDER BY night
            """;

        var rows = await connection.QueryAsync<UnavailableNightRow>(
            sql,
            new
            {
                listing_id = request.ListingId,
                check_in = request.Stay.CheckIn,
                check_out = request.Stay.CheckOut,
                cancellationToken,
            },
            transaction);

        return rows.Select(row => row.Night).ToArray();
    }
}
